flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1195,9 +1220,21 @@ private ApplicationReport startAppMaster(
         yarnClient.submitApplication(appContext);
 
         LOG.info("Waiting for the cluster to be allocated");
+
+        ApplicationReport report = waitForAppSubmitted(appId, false);
+
+        // since deployment was successful, remove the hook
+        ShutdownHookUtil.removeShutdownHook(deploymentFailureHook, getClass().getSimpleName(), LOG);
+        return report;
+    }
+
+    private ApplicationReport waitForAppSubmitted(
```

Comment:
Maybe it is better to create an `ApplicationReportProvider` that encapsulates the logic for how to retrieve the application report and until when to wait. Then the caller of `startAppMaster` can decide whether it wants to wait on it or not.

Reply:
Now `startAppMaster` will return the `ApplicationReportProvider`